File: testers/sqlserver/scala/src/scala/testdb/TestInsertTest.scala
package testdb

import org.junit.Assert._
import org.junit.Test
import testdb.userdefined.Email

import scala.util.Random

/** Tests for the TestInsert helper class that generates random test data. Tests seeded randomness, customization, and foreign key handling.
  */
class TestInsertTest {
  private val testInsert = TestInsert(Random(42))

  @Test
  def testCustomersInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Customers(email = Email("test@example.com"))
    assertNotNull(row)
    assertNotNull(row.customerId)
    assertNotNull(row.name)
  }

  @Test
  def testCustomersWithCustomization(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Customers(email = Email("custom@example.com"), name = "Custom Name")

    assertNotNull(row)
    assertEquals("Custom Name", row.name)
  }

  @Test
  def testProductsInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Products()

    assertNotNull(row)
    assertNotNull(row.productId)
    assertNotNull(row.name)
    assertNotNull(row.price)
  }

  @Test
  def testAllScalarTypesInsert(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.AllScalarTypes()

    assertNotNull(row)
    assertNotNull(row.id)
    assertNotNull(row.colRowversion)
  }

  @Test
  def testOrdersWithCustomerFK(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("orders-fk@example.com"))

    val order = testInsert.Orders(customer.customerId)

    assertNotNull(order)
    assertEquals(customer.customerId, order.customerId)
  }

  @Test
  def testOrderItemsWithFKs(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val customer = testInsert.Customers(email = Email("orderitems-fk@example.com"))
    val product = testInsert.Products()
    val order = testInsert.Orders(customer.customerId)

    val orderItem = testInsert.OrderItems(order.orderId, product.productId)

    assertNotNull(orderItem)
    assertEquals(order.orderId, orderItem.orderId)
    assertEquals(product.productId, orderItem.productId)
  }

  @Test
  def testMultipleInserts(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row1 = testInsert.Customers(email = Email("multi1@example.com"))
    val row2 = testInsert.Customers(email = Email("multi2@example.com"))
    val row3 = testInsert.Customers(email = Email("multi3@example.com"))

    assertNotEquals(row1.customerId, row2.customerId)
    assertNotEquals(row2.customerId, row3.customerId)
    assertNotEquals(row1.customerId, row3.customerId)
  }

  @Test
  def testInsertWithSeededRandom(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val testInsert1 = TestInsert(Random(123))
    val testInsert2 = TestInsert(Random(123))

    val row1 = testInsert1.Customers(email = Email("seeded1@test.com"))
    // Use different email to avoid UNIQUE constraint violation (same seed = same email)
    val row2 = testInsert2.Customers(email = Email("seeded2@test.com"))

    assertEquals(row1.name, row2.name)
  }

  @Test
  def testChainedCustomization(): Unit = withConnection { c =>
    given java.sql.Connection = c

    val row = testInsert.Customers(email = Email("first@test.com"), name = "First")

    assertEquals("First", row.name)
    assertEquals(Email("first@test.com"), row.email)
  }
}

File: testers/sqlserver/java/src/java/testdb/TestInsertTest.java
package testdb;

import static org.junit.Assert.*;

import java.util.Random;
import org.junit.Test;
import testdb.userdefined.Email;

/**
 * Tests for the TestInsert helper class that generates random test data. Tests seeded randomness,
 * customization, and foreign key handling.
 */
public class TestInsertTest {
  private final TestInsert testInsert = new TestInsert(new Random(42));

  @Test
  public void testCustomersInsert() {
    SqlServerTestHelper.run(
        c -> {
          var inserter = testInsert.Customers();

          var row = inserter.insert(c);
          assertNotNull(row);
          assertNotNull(row.customerId());
          assertNotNull(row.name());
        });
  }

  @Test
  public void testCustomersWithCustomization() {
    SqlServerTestHelper.run(
        c -> {
          var inserter = testInsert.Customers();

          var row = inserter.with(r -> r.withName("Custom Name")).insert(c);

          assertNotNull(row);
          assertEquals("Custom Name", row.name());
        });
  }

  @Test
  public void testProductsInsert() {
    SqlServerTestHelper.run(
        c -> {
          var inserter = testInsert.Products();
          var row = inserter.insert(c);

          assertNotNull(row);
          assertNotNull(row.productId());
          assertNotNull(row.name());
          assertNotNull(row.price());
        });
  }

  @Test
  public void testAllScalarTypesInsert() {
    SqlServerTestHelper.run(
        c -> {
          var row = testInsert.AllScalarTypes().insert(c);

          assertNotNull(row);
          assertNotNull(row.id());
          assertNotNull(row.colRowversion());
        });
  }

  @Test
  public void testOrdersWithCustomerFK() {
    SqlServerTestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);

          var order = testInsert.Orders(customer.customerId()).insert(c);

          assertNotNull(order);
          assertEquals(customer.customerId(), order.customerId());
        });
  }

  @Test
  public void testOrderItemsWithFKs() {
    SqlServerTestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var product = testInsert.Products().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);

          var orderItem = testInsert.OrderItems(order.orderId(), product.productId()).insert(c);

          assertNotNull(orderItem);
          assertEquals(order.orderId(), orderItem.orderId());
          assertEquals(product.productId(), orderItem.productId());
        });
  }

  @Test
  public void testMultipleInserts() {
    SqlServerTestHelper.run(
        c -> {
          var row1 = testInsert.Customers().insert(c);
          var row2 = testInsert.Customers().insert(c);
          var row3 = testInsert.Customers().insert(c);

          assertNotEquals(row1.customerId(), row2.customerId());
          assertNotEquals(row2.customerId(), row3.customerId());
          assertNotEquals(row1.customerId(), row3.customerId());
        });
  }

  @Test
  public void testChainedCustomization() {
    SqlServerTestHelper.run(
        c -> {
          var row =
              testInsert
                  .Customers()
                  .with(r -> r.withName("First"))
                  .with(r -> r.withEmail(new Email("first@test.com")))
                  .insert(c);

          assertEquals("First", row.name());
          assertEquals(new Email("first@test.com"), row.email());
        });
  }
}
